### ISSUE TITLE
[BUG] Failure to Evaluate 'Generate HCL' Configurations - Problem with Maps?

### ISSUE THREAD

**Issue by ndunn990** [mineiros-io/terramate#861]
# Description
I upgraded from `0.2.9` to `0.2.13`. Terramate suddenly seems unhappy with some of our 'generate_hcl' content blocks, but as far as I can tell they're perfectly valid. It's possible, though, that there's something we're doing incorrectly.

I have a few templates that are throwing the same error. I've noticed it always appears to struggle with a map.
They all say the same thing. As a very rough guess with little-to-no information, Terramate is no longer  evaluating maps properly. But again, that's mostly speculation on my part.

# Steps to Reproduce

Example Terramate template.
````
generate_hcl "_dns_vpc_associations.tf" {
  content {
    module "dns_vpc_associations_strateos" {
      source    = "../../../modules/aws-dns-vpc-associations"
      providers = {
        aws.vpc         = aws.primary
        aws.hosted_zone = aws.primary
      }
      vpc_associations = {
        private_default_aws_vpn = {
          vpc_id  = global.default_aws_vpn_client.vpc_id
          zone_id = module.dns.private_zone_id
        }
        private_internal_default_aws_vpn = {
          vpc_id  = global.default_aws_vpn_client.vpc_id
          zone_id = module.dns.private_internal_zone_id
        }
      }
    }
  }
}
````
Upon `terramate generate` command, I receive an error:
````
- /path/to/file/default_dns_vpc_associations.tm.hcl:5,19-8,8: evaluating content block: generate_hcl "_dns_vpc_associations.tf": partial evaluation failed: /path/to/file/default_dns_vpc_associations.tm.hcl:6,9-16: If this expression is intended to be a reference, wrap it in parentheses. If it's instead intended as a literal name containing periods, wrap it in quotes to create a string literal.
````

The error then disappears if I change my template as such:
````
generate_hcl "_dns_vpc_associations.tf" {
  content {
    module "dns_vpc_associations_strateos" {
      source    = "../../../modules/aws-dns-vpc-associations"
      providers = {
        (aws.vpc)         = aws.primary
        (aws.hosted_zone) = aws.primary
      }
      vpc_associations = {
        private_default_aws_vpn = {
          vpc_id  = global.default_aws_vpn_client.vpc_id
          zone_id = module.dns.private_zone_id
        }
        private_internal_default_aws_vpn = {
          vpc_id  = global.default_aws_vpn_client.vpc_id
          zone_id = module.dns.private_internal_zone_id
        }
      }
    }
  }
}
````
Another example:
````
generate_hcl "_main.tf" {
  lets {
    default_node_group_name = tm_try(global.eks.node_groups.default_group_name, "${global.env}-prim-default")
  }
  content {
    module "eks_cluster" {
      source = global.module_sources.eks.source
      version = global.module_sources.eks.version

      providers = { aws = aws.primary }
      eks_managed_node_groups = merge({ let.default_node_group_name = local.node_groups.default }, local.node_groups.additional)
      # More Terraform that is not relevant here...
    }
  }
}
````

I see the exact same error. I can fix it by changing this line and adding parentheses around `let.default_node_group_name`: 
````
eks_managed_node_groups = merge({ (let.default_node_group_name) = local.node_groups.default }, local.node_groups.additional)
````

# Expected Behavior
I expected Terramate to properly parse my `generate_hcl` configuration and generate the Terraform code as previous versions of Terramate did.

# Environment
 - OS.: Linux
 - OS Version: Ubuntu 22.04
 - Git Version: 2.34.1
 - Terramate Version: 0.2.13

**Comment by i4ki** [comment]
@ndunn990 As @mariux said, Terramate now is behaving like default HCL evaluator. See here:
https://go.dev/play/p/wUb09u5E7ls

But this doesn't combine well when generating code targeting Terraform... We will fix this.
Thanks

**Comment by mariux** [comment]
We just had an internal meeting about this topic and decided:

After releasing a fix we will also provide a clearer interface for those cases in an upcoming release. 

So the following will be available in `generate_hcl` soon, assuming the stack id is "deadbeef":

```
(terramate.stack.id)  = 1   # generates => "deadbeef" = 1
terramate.stack.id    = 1   # generates => terramate.stack.id = 1
"terramate.stack.id"  = 1   # generates => "terramate.stack.id" = 1
(unknown.whatever)    = 1   # generates => (unknown.whatever) = 1
```

We will also check if it makes sense to extend this behavior into `globals` later on.

any feedback is welcome here ;)

**Comment by ndunn990** [comment]
> We just had an internal meeting about this topic and decided:
> 
> After releasing a fix we will also provide a clearer interface for those cases in an upcoming release.
> 
> So the following will be available in `generate_hcl` soon, assuming the stack id is "deadbeef":
> 
> ```
> (terramate.stack.id)  = 1   # generates => "deadbeef" = 1
> terramate.stack.id    = 1   # generates => terramate.stack.id = 1
> "terramate.stack.id"  = 1   # generates => "terramate.stack.id" = 1
> (unknown.whatever)    = 1   # generates => (unknown.whatever) = 1
> ```
> 
> We will also check if it makes sense to extend this behavior into `globals` later on.
> 
> any feedback is welcome here ;)

I believe this makes sense to me. It preserves previous functionality, while providing more options/flexibility for how one might want to generate HCL/Terraform code. I also don't think this will be difficult to adapt to - it should just require a few syntax changes to some of our existing 'generate_hcl' templates.

I appreciate all of you looking into this. I downgraded my version of Terramate in the meantime and will keep an eye out for the next release.

**Comment by mariux** [comment]
@ndunn990 as a first fix, could you test the following branch? `[i4k-fix-object-keys-eval](https://github.com/mineiros-io/terramate/tree/i4k-fix-object-keys-eval)`

you can install a branch using instructions from our [README](https://github.com/mineiros-io/terramate#readme): e.g. `go install github.com/mineiros-io/terramate/cmd/terramate@i4k-fix-object-keys-eval`

It does not yet implement the behavior shared in the last comment, but should fix the regression.

We are planning to merge and release it on Monday latest.

**Comment by Xcess** [comment]
In case someone's wondering how to downgrade to 0.2.12 on MacOS:

```
brew unlink terramate
cd /tmp
curl https://raw.githubusercontent.com/Homebrew/homebrew-core/72d1136bf7e357ae06d0423da405d4727eef2251/Formula/terramate.rb -O
brew install ./terramate.rb
```

**Comment by ndunn990** [comment]
> @ndunn990 as a first fix, could you test the following branch? [`i4k-fix-object-keys-eval`](https://github.com/mineiros-io/terramate/tree/i4k-fix-object-keys-eval) of #862
> 
> you can install a branch using instructions from our [README](https://github.com/mineiros-io/terramate#readme): e.g. `go install github.com/mineiros-io/terramate/cmd/terramate@i4k-fix-object-keys-eval`
> 
> It does not yet implement the behavior shared in the last comment, but should fix the regression.
> 
> We are planning to merge and release it on Monday latest.

I installed the version requested and I was able to run `terramate generate` with no errors. :+1: 
```
> terramate --version
0.2.14-dev
````

Just as a heads-up (especially for anyone reading through this thread in the future), I _did_ still notice some changes to our generated code with this newer version of Terramate. However, I believe all of the changes are  formatting and inconsequential. For instance, here's a slightly-sanitized portion of a `git diff`.
```
diff --git a/path/to/terraform/code.tf
index b721aa7..2318b15 100644
--- a/path/to/terraform/code.tf
+++ b/path/to/terraform/code.tf
@@ -2,7 +2,6 @@
 
 module "audit" {
   additional_project_source_repos = [
-    # Configuration Repository
     "ssh://git@github.com/our-org/our-repo.git",
   ]
   depends_on = [
@@ -11,37 +10,38 @@ module "audit" {
   environment         = "my-environment"
   project_description = "ArgoCD project."
   project_name        = "my-project"
-  providers           = { argocd = argocd.decentralized }
-  source              = "../../../../terraform-ng/modules/argocd_project"
+  providers = {
+    argocd = argocd.decentralized
+  }
+  source = "../path/to/modules/argocd_project"
 }
```
